### ISSUE TITLE
raco pkg install off non-"master" git branch: "git: could not find requested reference"

### ISSUE THREAD

**Issue by tonyg** [racket/racket#3973]
**What version of Racket are you using?**
```
root@tmp:/usr/racket/bin# ./racket --version
Welcome to Racket v8.2 [cs].
```

**What program did you run?**
```raco pkg install preserves```

**What should have happened?**
```
root@tmp:/usr/racket/bin# ./raco pkg install 'https://gitlab.com/preserves/preserves.git?path=implementations%2Fracket%2Fpreserves#main'Querying Git references for preserves at https://gitlab.com/preserves/preserves.git?path=implementations%2Fracket%2Fpreserves#main
Using cached16298772001629877200562 for https://gitlab.com/preserves/preserves.git?path=implementations%2Fracket%2Fpreserves#main
raco setup: version: 8.2
raco setup: platform: x86_64-linux [cs]
raco setup: target machine: ta6le
raco setup: installation name: 8.2
raco setup: variants: cs
raco setup: main collects: /usr/racket/collects/
raco setup: collects paths: 
raco setup:   /root/.local/share/racket/8.2/collects
raco setup:   /usr/racket/collects/
raco setup: main pkgs: /usr/racket/share/pkgs
raco setup: pkgs paths: 
raco setup:   /usr/racket/share/pkgs
raco setup:   /root/.local/share/racket/8.2/pkgs
raco setup: links files: 
raco setup:   /usr/racket/share/links.rktd
raco setup:   /root/.local/share/racket/8.2/links.rktd
raco setup: main docs: /usr/racket/doc
raco setup: --- updating info-domain tables ---                    [7:46:09]
raco setup: updating: /root/.local/share/racket/8.2/share/info-cache.rktd
raco setup: --- pre-installing collections ---                     [7:46:09]
raco setup: --- installing foreign libraries ---                   [7:46:09]
raco setup: --- installing shared files ---                        [7:46:09]
raco setup: --- compiling collections ---                          [7:46:09]
raco setup: making: <pkgs>/preserves/preserves
raco setup:  in <pkgs>/preserves/preserves
raco setup: making: <pkgs>/preserves/preserves-schema
raco setup:  in <pkgs>/preserves/preserves-schema
raco setup:  in <pkgs>/preserves/preserves-schema/gen
raco setup:  in <pkgs>/preserves/preserves-schema/bin
raco setup: making: <pkgs>/preserves/preserves-schema/bin
raco setup: making: <pkgs>/preserves/preserves-schema/gen
raco setup: making: <pkgs>/preserves/preserves/tests
raco setup:  in <pkgs>/preserves/preserves/tests
raco setup: --- creating launchers ---                             [7:46:12]
raco setup: launcher: /root/.local/share/racket/8.2/bin/preserves-tool-rkt
raco setup: launcher: /root/.local/share/racket/8.2/bin/preserves-schema-rkt
raco setup: --- installing man pages ---                           [7:46:12]
raco setup: --- building documentation ---                         [7:46:12]
raco setup: --- installing collections ---                         [7:46:13]
raco setup: --- post-installing collections ---                    [7:46:13]
```

**If you got an error message, please include it here.**
```
root@tmp:/usr/racket/bin# ./raco pkg install preserves
Resolving "preserves" via https://download.racket-lang.org/releases/8.2/catalog/
Resolving "preserves" via https://pkgs.racket-lang.org
Downloading repository https://gitlab.com/preserves/preserves.git?path=implementations%2Fracket%2Fpreserves#main
git: could not find requested reference
  reference: master
  repo: preserves/preserves.git
  context...:
   /usr/racket/collects/net/git-checkout.rkt:403:0: select-commits
   [repeats 1 more time]
   /usr/racket/collects/net/git-checkout.rkt:73:8
   /usr/racket/collects/net/git-checkout.rkt:54:2: retry-loop
   /usr/racket/collects/pkg/private/download.rkt:102:2: download!
   /usr/racket/collects/file/cache.rkt:63:2: fetch-and-continue
   /usr/racket/collects/racket/contract/private/arrow-val-first.rkt:555:3
   /usr/racket/collects/pkg/private/download.rkt:94:0: download-repo!
   /usr/racket/collects/pkg/private/stage.rkt:318:11
   /usr/racket/collects/pkg/private/stage.rkt:114:0: stage-package/info
   [repeats 1 more time]
   /usr/racket/collects/pkg/private/install.rkt:141:0: install-packages
   /usr/racket/collects/pkg/private/install.rkt:925:4
   /usr/racket/collects/racket/contract/private/arrow-val-first.rkt:555:3
   /usr/racket/collects/racket/file.rkt:437:8
   /usr/racket/collects/racket/file.rkt:426:0: call-with-file-lock
   ...
```

**Please include any other relevant details**
e.g., the operating system used or how you are running the code.

What's weird about this is that literally cut-and-pasting the URL from the error message and using that instead causes it to work out fine!

**Comment by mflatt** [comment]
I think you're seeing a bug in v8.2 that has been fixed for the next release. Meanwhile, I expect that it magically appears to be fixed even if you try v8.2 again, right now, but the bug will seem to reappear if you push to the repo again.

The problem is that when the pkg server is not up-to-date with the latest commit in the repo, then then `net/git-checkout` library fails to find the commit reported from the pkg server (i.e., not the latest commit) for a non-`master` branch. When you use the URL directly instead of going through the package server, then that mismatch can't happen.

This has been fixed for the next release by 06d892f477f8cb2899f5f41b57e01cef13dd9fe3.

**Comment by tonyg** [comment]
Thank you Matthew, that sounds very plausible! And indeed it works now that the pkg server has caught up.
